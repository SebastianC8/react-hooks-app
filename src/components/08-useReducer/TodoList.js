import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul>
            {
                todos.map((row, ix) => <TodoListItem key = { row.id }  todo = { row } index = { ix } handleDelete = { handleDelete } handleToggle = { handleToggle } /> )
            }
        </ul>
    )
}
