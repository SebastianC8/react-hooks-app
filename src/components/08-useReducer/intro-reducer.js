const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const toDoReducer = (state = initialState, action) => {

    switch (action?.type) {

        case 'add':
            return [...state, action.payload];
        break;

    }

    return state;
}

const newToDo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
}

const actionAdd = {
    type: 'add',
    payload: newToDo
}

const tareas = toDoReducer(initialState, actionAdd);
console.log(tareas);