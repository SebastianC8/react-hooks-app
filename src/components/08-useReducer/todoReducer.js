export const todoReducer = (state = [], action) => {

    switch (action.type) {

        case 'add':
            return [ ...state, action.payload ];

        case 'delete':
            return state.filter((row) => row.id !== action.payload);

        case 'toggle':
            return state.map((row) => (row.id === action.payload) ? { ...row, done: !row.done } : row );
    
        default:
            return state;
    }

}