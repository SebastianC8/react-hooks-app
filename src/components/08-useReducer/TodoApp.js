import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    const handleAddTodo = (newTodo) => {
        dispatch({ type: 'add', payload: newTodo });
    }

    const handleDelete = (id) => {
        const action = { type: 'delete', payload: id }
        dispatch(action);
    }

    const handleToggle = (id) => {
        const action = { type: 'toggle', payload: id }
        dispatch(action);
    }

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    return (
        <div>
            <h1>TodoAPP ({ todos.length })</h1>
            <hr/>

            <div className="row">

                <div className="col-7">
                    <TodoList todos = { todos } handleDelete = { handleDelete } handleToggle = { handleToggle } />
                </div>

                <div className="col-5">
                    <TodoAdd handleAddTodo = { handleAddTodo }/>
                </div>
                
            </div>
            
        </div>
    )
}
