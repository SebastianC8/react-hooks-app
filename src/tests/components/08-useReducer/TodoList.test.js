import { React } from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';

describe('Pruebas en TodoList', () => {

    const demoTodos = [
        { id: 1, desc: 'Aprender React', done : false },
        { id: 2, desc: 'Aprender Python', done : false }
    ];

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList
            todos = { demoTodos }
            handleDelete = { handleDelete }
            handleToggle = { handleToggle }
        />
    );

    test('Show ok', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('Debe tener 2 <TodoListItem/>', () => {

        expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
        expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));

    });

});