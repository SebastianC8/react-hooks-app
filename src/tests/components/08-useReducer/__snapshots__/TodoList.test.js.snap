// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoList Show ok 1`] = `
<ul>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Python",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
